### PR TITLE
Refactor how we handle and validate LN ConnectionStrings

### PR DESCRIPTION
**Pull request by NicolasDorier**
This PR is refactoring how we handle and validate connection strings for LN payment methods.

Because connection string can read the file system, we must be sure that a malicious party can't leverage lightning connection string to read and steal arbitrary file on the host.

While the original code was doing this successfully, it was hard to read and hard to understand.

I simplified the code in the following way: Now a Lightning Payment Method can be either an external connection string (as before), but some fields forbidden (such as macaroonfilepath).
If the user wants to use the internal node, then the connection string should be the string "Internal Node".

The nice thing about it is that if the user change his internal lightning implementation, he won't need to re-set the connection string to the new internal node.

I am also migrating all stores which used the internal node before to the new connection string "Internal Node".

long story short:
1. The connection strings are now strictly limited to a subset if you are not admin "IsSafe()"
2. For using internal node, you need to set the connection string to "Internal Node"
3. A migration has been done so connection strings that match internal node are converted to "Internal Node"
4. Some cleanup of old fields in the DB that we don't use anymore
5. The new design means that it is possible to set lightning network's payment method to internal in the greenfield API @Kukks

### CODE REVIEW

**Comment by Kixunil** [comment]
This sounds reasonable, good point about reading arbitrary files!

I'd still love to see the ability to have multiple internal nodes. It doesn't need to be implemented right away. Maybe to make it easier connection string could be `InternalNode=LABEL`? The initial implementation can use fixed `InternalNode=main` value.

**Comment by NicolasDorier** [comment]
That's good idea, after this refactoring it can be easily implemented later.

File: BTCPayServer/Views/Stores/AddLightningNode.cshtml [hunk 1]
```diff
@@ -42,6 +42,14 @@
             <div class="form-group">
                 <p class="mb-2">The connection string encapsulates the configuration for connecting to your lightning node. BTCPay Server currently supports:</p>
                 <ul>
+                    <li class="mb-2">
+                        <strong>Internal node</strong>, if you are administrator of the server:
+                        <ul>
+                            <li>
+                                <code>Internal Node</code>
```
Comment (Kukks) [inline BTCPayServer/Views/Stores/AddLightningNode.cshtml:49]: Is it supposed to be with a space here too?

File: BTCPayServer/Payments/Lightning/LightningListener.cs [hunk 2]
```diff
@@ -376,15 +394,26 @@ async Task Listen(CancellationToken cancellation)
...
+                Logs.PayServer.LogError(ex, $"{supportedPaymentMethod.CryptoCode} (Lightning): Error while contacting {GetLightningUrl().BaseUri}");
+                Logs.PayServer.LogInformation($"{supportedPaymentMethod.CryptoCode} (Lightning): Stop listening {GetLightningUrl().BaseUri}");
             }
             catch (OperationCanceledException) when (cancellation.IsCancellationRequested) { }
             if (_ListenedInvoices.IsEmpty)
-                Logs.PayServer.LogInformation($"{supportedPaymentMethod.CryptoCode} (Lightning): No more invoice to listen on {supportedPaymentMethod.GetLightningUrl().BaseUri}, releasing the connection.");
+                Logs.PayServer.LogInformation($"{supportedPaymentMethod.CryptoCode} (Lightning): No more invoice to listen on {GetLightningUrl().BaseUri}, releasing the connection.");
+        }
+
+        private LightningConnectionString GetLightningUrl()
```
Comment (Kukks) [inline BTCPayServer/Payments/Lightning/LightningListener.cs:405]: Duplicated code (add as static helper somewhere?)

File: BTCPayServer/Payments/Lightning/LightningListener.cs [hunk 3]
```diff
@@ -240,6 +244,16 @@ private async Task CreateNewLNInvoiceForBTCPayInvoice(InvoiceEntity invoice)
 
         }
 
+        private LightningConnectionString GetLightningUrl(LightningSupportedPaymentMethod supportedMethod)
```
Comment (Kukks) [inline BTCPayServer/Payments/Lightning/LightningListener.cs:247]: Are you sure this is correct? 
It seems to try and see if payment method has a value and if not sets it to internal node.
What if a user finds a way to empty the conn string without deleting the method (non admin would gain access to internal node)?

File: BTCPayServer/Payments/Lightning/LightningLikePaymentHandler.cs [hunk 4]
```diff
@@ -135,6 +142,21 @@ public async Task<NodeInfo> GetNodeInfo(bool preferOnion, LightningSupportedPaym
             }
         }
 
+        private ILightningClient CreateLightningClient(LightningSupportedPaymentMethod supportedPaymentMethod, BTCPayNetwork network)
```
Comment (Kukks) [inline BTCPayServer/Payments/Lightning/LightningLikePaymentHandler.cs:145]: duplicated method

File: BTCPayServer/Controllers/StoresController.LightningLike.cs [hunk 5]
```diff
@@ -170,10 +143,9 @@ public async Task<IActionResult> AddLightningNode(string storeId, LightningNodeV
                     return View(vm);
             }
         }
-
         private bool CanUseInternalLightning()
         {
-            return (_BTCPayEnv.IsDeveloping || User.IsInRole(Roles.ServerAdmin) || _CssThemeManager.AllowLightningInternalNodeForAll);
+            return User.IsInRole(Roles.ServerAdmin) || _CssThemeManager.AllowLightningInternalNodeForAll;
```
Comment (Kukks) [inline BTCPayServer/Controllers/StoresController.LightningLike.cs:148]: This might break tests?

File: BTCPayServer/Controllers/StoresController.LightningLike.cs [hunk 6]
```diff
@@ -88,40 +92,9 @@ public async Task<IActionResult> AddLightningNode(string storeId, LightningNodeV
...
-                    {
-                        ModelState.AddModelError(nameof(vm.ConnectionString), "The macaroonfilepath should be fully rooted");
-                        return View(vm);
-                    }
-                }
-
-                if (isInternalNode && !CanUseInternalLightning())
-                {
-                    ModelState.AddModelError(nameof(vm.ConnectionString), "Unauthorized url");
+                    ModelState.AddModelError(nameof(vm.ConnectionString), $"You are not admin, so the connection string should not contains 'cookiefilepath', 'macaroondirectorypath', 'macaroonfilepath', and should not point to a local ip or to a dns name ending with '.internal', '.local', '.lan' or '.'.");
```
Comment (Kukks) [inline BTCPayServer/Controllers/StoresController.LightningLike.cs:97]:
```suggestion
                    ModelState.AddModelError(nameof(vm.ConnectionString), $"You are not a server admin, so the connection string should not contain 'cookiefilepath', 'macaroondirectorypath', 'macaroonfilepath', and should not point to a local ip or to a dns name ending with '.internal', '.local', '.lan' or '.'.");
```

File: BTCPayServer.Tests/GreenfieldAPITests.cs [hunk 7]
```diff
@@ -1287,33 +1288,85 @@ public async Task LightningNetworkPaymentMethodAPITests()
...
             {
-                await client.GetStoreOnChainPaymentMethod(store.Id, "BTC");
+                await adminClient.GetStoreOnChainPaymentMethod(store.Id, "BTC");
+            });
+
+
+            // Let's verify that the admin client can't change LN to unsafe connection strings without modify server settings rights
+            foreach (var forbidden in new string[]
+            {
+                "type=clightning;server=tcp://127.0.0.1",
```
Comment (Kukks) [inline BTCPayServer.Tests/GreenfieldAPITests.cs:1327]: what about 0.0.0.0, and even the ip of the server?

Comment (Kukks) [reply BTCPayServer.Tests/GreenfieldAPITests.cs:1327]: or even ipv6 local ones?

File: BTCPayServer/Hosting/MigrationStartupTask.cs [hunk 8]
```diff
@@ -114,6 +128,94 @@ public async Task ExecuteAsync(CancellationToken cancellationToken = default)
...
+                    });
+                    store.DerivationStrategies = jObject.ToString();
+                    store.DerivationStrategy = null;
+                }
+                if (string.IsNullOrEmpty(store.DerivationStrategies))
+                    continue;
+
+                var strats = JObject.Parse(store.DerivationStrategies);
+                bool updated = false;
+                foreach (var prop in strats.Properties().Where(p => p.Name.EndsWith("LightningLike", StringComparison.OrdinalIgnoreCase)))
```
Comment (Kukks) [inline BTCPayServer/Hosting/MigrationStartupTask.cs:163]: Should try and parse with PaymentMethodId to be safe

File: BTCPayServer/Hosting/MigrationStartupTask.cs [hunk 9]
```diff
@@ -114,6 +128,94 @@ public async Task ExecuteAsync(CancellationToken cancellationToken = default)
...
+                {
+                    var method = ((JObject)prop.Value);
+                    var lightningCharge = method.Property("LightningChargeUrl", StringComparison.OrdinalIgnoreCase);
+                    var ln = method.Property("LightningConnectionString", StringComparison.OrdinalIgnoreCase);
+                    if (lightningCharge != null)
```
Comment (Kukks) [inline BTCPayServer/Hosting/MigrationStartupTask.cs:168]: why is charge so special?

**Review by Kukks — CHANGES_REQUESTED** [review]
simplifies and cleans up all the hardcore logic around internal node checks

File: BTCPayServer/Controllers/StoresController.LightningLike.cs (hunk 5 above)
Comment (NicolasDorier) [reply BTCPayServer/Controllers/StoresController.LightningLike.cs:148]: I changed tests so the user is admin.

File: BTCPayServer.Tests/GreenfieldAPITests.cs (hunk 7 above)
Comment (NicolasDorier) [reply BTCPayServer.Tests/GreenfieldAPITests.cs:1327]: @Kukks yeah, both should be forbidden, I can add test for that.

File: BTCPayServer/Hosting/MigrationStartupTask.cs (hunk 8 above)
Comment (NicolasDorier) [reply BTCPayServer/Hosting/MigrationStartupTask.cs:163]: I was thinking about it, but then we have a problem if somebody disabled shitcoins. (As payment id will not be parseable)

File: BTCPayServer/Hosting/MigrationStartupTask.cs (hunk 9 above)
Comment (NicolasDorier) [reply BTCPayServer/Hosting/MigrationStartupTask.cs:168]: Charge was the first lightning implementation supported by BTCPay, so before there was 3 fields Username/Password/ChargeUrl rather than a single ConnectionString.

We probably changed that 1 or 2 years ago, but those fields were still there.

File: BTCPayServer/Payments/Lightning/LightningListener.cs (hunk 3 above)
Comment (NicolasDorier) [reply BTCPayServer/Payments/Lightning/LightningListener.cs:247]: not possible, setting the connection string is only done through `void SetLightningUrl(LightningConnectionString connectionString)` which verify that: `connectionString` is not null.
`LightningConnectionString` make it sure it is not empty.

File: BTCPayServer/Views/Stores/AddLightningNode.cshtml (hunk 1 above)
Comment (NicolasDorier) [reply BTCPayServer/Views/Stores/AddLightningNode.cshtml:49]: yes

**Comment by Kixunil** [comment]
Since this changes connection string validation, I wonder how this affects #1603 
If it doesn't, would it make sense to bundle such change here?

**Comment by NicolasDorier** [comment]
It probably does, will test if I can repro. I don't check http/https anymore in this PR. I just check that the external configuration string does not contains local stuff, so the checks are now simpler.

File: BTCPayServer/Payments/Lightning/LightningListener.cs (hunk 3 above)
Comment (NicolasDorier) [reply BTCPayServer/Payments/Lightning/LightningListener.cs:247]: I will put a separated string instead. Make it harder to shoot in the foot.

File: BTCPayServer/Views/Stores/AddLightningNode.cshtml (hunk 1 above)
Comment (Kixunil) [reply BTCPayServer/Views/Stores/AddLightningNode.cshtml:49]: I'm afraid this will lead to confusion if other ways are without space. Do you have some particular reason for it?

**Review by Kukks — APPROVED** [review]
ACCK
